apps/prairielearn/python/traverse.py

```diff
@@ -0,0 +1,153 @@
+from collections import deque
+from html import escape as html_escape
+from itertools import chain
+from typing import Callable, Deque, List, Optional, Tuple, Union
+
+import lxml.html
+
+ElementReplacement = Optional[
+    Union[str, lxml.html.HtmlElement, List[lxml.html.HtmlElement]]
+]
+
+# https://developer.mozilla.org/en-US/docs/Glossary/Void_element
+VOID_ELEMENTS = frozenset(
+    {
+        "area",
+        "base",
+        "br",
+        "col",
+        "embed",
+        "hr",
+        "img",
+        "input",
+        "link",
+        "meta",
+        "source",
+        "track",
+        "wbr",
+    }
+)
+
+
+def traverse_and_execute(
+    html: str, fn: Callable[[lxml.html.HtmlElement], None]
+) -> None:
+    elements = lxml.html.fragments_fromstring(html)
+
+    # If there's leading text, the first element of the array will be a string.
+    # We can just discard that.
+    if isinstance(elements[0], str):
+        del elements[0]
+
+    for e in chain.from_iterable(element.iter() for element in elements):
+        fn(e)
+
+
+def format_attrib_value(v: str) -> str:
+    # https://html.spec.whatwg.org/multipage/parsing.html#escapingString
+    return v.replace("&", "&amp;").replace('"', "&quot;").replace("\xa0", "&nbsp;")
+
+
+def get_source_definition(element: lxml.html.HtmlElement) -> str:
+    attributes = (
+        f'''{k}="{format_attrib_value(v)}"''' for k, v in element.attrib.items()
+    )
+    return f"<{' '.join((element.tag, *attributes))}>"
+
+
+def traverse_and_replace(
+    html: str, replace: Callable[[lxml.html.HtmlElement], ElementReplacement]
+) -> str:
+    # Initialize result and work data structures
+    result: Deque[str] = deque()
+
+    initial_list = lxml.html.fragments_fromstring(html)
+    count_stack: Deque[int] = deque([len(initial_list)])
```

Comment:
Why do we need the `count_stack`? I'm not really following how this code works. Maybe a brief overview in a comment would help?

Reply:
Basically, the traversal algorithm needs this to track where to insert closing tags (since a single element can have multiple direct children). If you think of the HTML that's being parsed as a tree, this count stack just tracks where in that tree the algorithm is currently processing.

This algorithm is actually really similar to one solving the [balanced brackets](https://medium.com/@eric_lum/the-balanced-brackets-problem-ffc01acee8ee) problem, with the main difference being a slight change in how closing brackets (in this case closing tags) are handled.

Reply:
cc @eliotwrobson want to add a description for your code here?

Reply:
@nwalters512 a brief description suggested here https://github.com/PrairieLearn/PrairieLearn/pull/7128/files#r1222267135